### DISCUSSION TITLE
Example/Documentation/Guides for using spaCy cython classes and structs

### DISCUSSION THREAD

**Discussion by narayanacharya6** [explosion/spaCy#9427] [Help: Coding & Implementations] [answered]
Can someone point to any examples or documentation for getting started with using classes and structs mentioned in the [Cython section of the documentation.](https://spacy.io/api/cython-classes). I am new to using Cython and should probably read more about it but wanted to implement something quickly and hence need some guidance.

What I have tried:

1. I tried to cimport the Doc class in a file, say `sample.py`
```
import numpy as np
cimport numpy as np
from spacy.tokens import Doc
from spacy.tokens cimport Doc
```
Pycharm complains on the last line but `python setup.py build_ext --inplace` compiles the file fine and produces a `sample.cpp` file. Note that it does NOT complain about the `cimport numpy as np`

2. When I try to use the file it complains `ModuleNotFoundError: No module named 'spacy.tokens.Doc'`

Do I need to setup something more in order to get started with this? If so, link to an example or guide will be really appreciated :)

**Comment by narayanacharya6** [comment, marked as answer]
Issue mentioned above was solved. The spaCy directory was not included in the `include_dirs` param of the `Extension` in my `setup.py`.  

Follow up question would be if spaCy some method to return directory where to locate the header files similar to `numpy.get_include()`

Still looking for examples/guides for using spaCy cython classes and structs ;) Especially those that have some iterations over doc tokens and sentence spans.

**Reply by polm** [reply]
Sorry, but right now the best place to look is just at the source. The Cython API is kind of do-it-yourself really. We are working on dev docs in general and plan to cover the Cython API as part of that, but we don't have any particular schedule for that so it may be a while. 

Part of this is that while the Cython API is intended to be externally usable, we haven't made use of it that way and I don't think there are many libraries that do it either, so we haven't had a lot of feedback on it from that perspective so far, though we would welcome it.

> Follow up question would be if spaCy some method to return directory where to locate the header files similar to numpy.get_include()

There is no feature like that, no.
